server/src/test/java/org/elasticsearch/search/source/SourceFetchingIT.java

```diff
@@ -95,4 +99,33 @@ public void testSourceWithWildcardFiltering() {
         assertThat(response.getHits().getAt(0).getSourceAsMap().size(), equalTo(1));
         assertThat((String) response.getHits().getAt(0).getSourceAsMap().get("field"), equalTo("value"));
     }
+
+    public void testSourceFilteringWithJsonField() {
+        prepareCreate("test").addMapping("_doc", "headers", "type=json").get();
+        ensureGreen();
+
+        Map<String, Object> headers = new HashMap<>();
+        headers.put("content-type", "application/json");
+        headers.put("origin", "https://www.elastic.co");
+        Map<String, Object> source = Collections.singletonMap("headers", headers);
+
+        client().prepareIndex("test", "_doc", "1").setSource(source).get();
+        refresh();
+
+        SearchResponse response = client().prepareSearch("test").setFetchSource(true).get();
+        assertThat(response.getHits().getAt(0).getSourceAsMap(), equalTo(source));
+
+        response = client().prepareSearch("test").setFetchSource("headers", null).get();
+        assertThat(response.getHits().getAt(0).getSourceAsMap(), equalTo(source));
+
+        response = client().prepareSearch("test").setFetchSource("headers.content-type", null).get();
+        Map<String, Object> filteredSource = Collections.singletonMap("headers",
+            Collections.singletonMap("content-type", "application/json"));
+        assertThat(response.getHits().getAt(0).getSourceAsMap(), equalTo(filteredSource));
+
+        response = client().prepareSearch("test").setFetchSource(null, "headers.content-type").get();
```

Comment:
Nit: Could you add a comment to this block and the block above explaining that this is testing the exclude filtering and the one above is testing the include filtering?

Reply:
👍 I don't think about this much these days because IntelliJ shows parameter name hints :)